include/executors/reduction.hpp

```diff
@@ -0,0 +1,132 @@
+    } else {
+      // THE OTHER CASES ALWAYS USE THE BINARY FUNCTION
+      auto localExpr = ReductionExpr<oper_type, LHS_type, LHS_type>(
+              ((nWG == 1) ? lhs : (even ? opShMem2 : opShMem1)),
+              (even ? opShMem1 : opShMem2), localSize, globalSize);
+      execute_subtree<Device, using_shared_mem::enabled>(dev.sycl_queue(), Evaluator<ExpressionT>(localExpr), localSize, globalSize, sharedSize);
+    }
+    _N = nWG;
+    nWG = (_N + (2 * localSize) - 1) / (2 * localSize);
+    frst = false;
+    even = !even;
+  } while (_N > 1);
+};
+
+/*!
+ * @brief Applies a reduction to a tree, receiving a scratch buffer.
+ */
+/* template <typename EvaluatorT, typename Scratch> */
+/* void reduce(SYCLDevice &dev, EvaluatorT ev, Scratch scr) { */
+/*   using oper_type = typename blas::Converter<Evaluator<ExpressionT, Device>>::oper_type; */
+/*   using input_type = typename blas::Converter<Evaluator<ExpressionT, Device>>::input_type; */
+/*   using cont_type = typename blas::Converter<Evaluator<ExpressionT, Device>>::cont_type; */
+/*   using LHS_type = typename blas::Converter<Evaluator<ExpressionT, Device>>::LHS_type; */
+/*   auto _N = ev.getSize(); */
+/*   auto localSize = ev.blqS; */
+/*   // IF THERE ARE ENOUGH ELEMENTS, EACH BLOCK PROCESS TWO BLOCKS OF ELEMENTS */
+/*   // THEREFORE, 2*GLOBALSIZE ELEMENTS ARE PROCESSED IN A STEP */
+/*   // MOREOVER, A LOOP ALLOWS TO REPEAT THE PROCESS UNTIL */
+/*   // ALL THE ELEMENTS ARE PROCESSED */
+/*   auto nWG = (ev.grdS + (2 * localSize) - 1) / (2 * localSize); */
+/*   auto lhs = ev.l; */
+/*   auto rhs = ev.r; */
+
+/*   // Two accessors to local memory */
+/*   auto sharedSize = ((nWG < localSize) ? localSize : nWG); */
+/*   auto opShMem1 = LHS_type(scr, 0, 1, sharedSize); */
+/*   auto opShMem2 = LHS_type(scr, sharedSize, 1, sharedSize); */
+
+/*   bool frst = true; */
+/*   bool even = false; */
+/*   do { */
+/*     auto globalSize = nWG * localSize; */
+/*     if (frst) { */
+/*       // THE FIRST CASE USES THE ORIGINAL BINARY/TERNARY FUNCTION */
+/*       auto localEval = input_type(((nWG == 1) ? lhs : opShMem1), rhs, localSize, globalSize); */
+/*       execute_subtree<Device, using_shared_mem::enabled>(dev.sycl_queue(), localExpr, localSize, */
+/*                                               globalSize, sharedSize); */
+/*     } else { */
+/*       // THE OTHER CASES ALWAYS USE THE BINARY FUNCTION */
+/*       auto localEval = Evaluator<ReductionExpr<oper_type, LHS_type, LHS_type>, Device>( */
+/*               ((nWG == 1) ? lhs : (even ? opShMem2 : opShMem1)), */
+/*               (even ? opShMem1 : opShMem2), localSize, globalSize); */
+/*       execute_subtree<Device, using_shared_mem::enabled>(dev.sycl_queue(), localExpr, localSize, globalSize, sharedSize); */
+/*     } */
+/*     _N = nWG; */
+/*     nWG = (_N + (2 * localSize) - 1) / (2 * localSize); */
+/*     frst = false; */
+/*     even = !even; */
+/*   } while (_N > 1); */
+/* } */
```

Comment:
We discussed it with @mehdi-goli and have not found any place this function is used.